ord_schema/process_dataset.py

```diff
@@ -122,32 +127,72 @@ def _update_reaction(reaction):
         reaction.provenance.record_id = record_id
 
 
+def _get_filenames():
+    """Gets a list of Dataset proto filenames to process.
+
+    Returns:
+        List of text filenames.
+    """
+    if FLAGS.input_pattern and FLAGS.input_file:
+        raise ValueError(
+            'one of --input_pattern or --input_file is required, not both')
+    if FLAGS.input_pattern:
+        # Setting recursive=True allows recursive matching with '**'.
+        return glob.glob(FLAGS.input_pattern, recursive=True)
+    if FLAGS.input_file:
+        with open(FLAGS.input_file) as f:
+            return [line.strip() for line in f]
+    raise ValueError('one of --input_pattern or --input_file is required')
+
+
+# pylint: disable=too-many-branches
 def main(argv):
     del argv  # Only used by app.run().
-    # Setting recursive=True allows recursive matching with '**'.
-    filenames = glob.glob(FLAGS.input_pattern, recursive=True)
+    filenames = sorted(_get_filenames())
+    if not filenames:
+        logging.info('nothing to do')
+        return  # Nothing to do.
     datasets = {}
     for filename in filenames:
         datasets[filename] = message_helpers.load_message(
             filename, dataset_pb2.Dataset, FLAGS.input_format)
     if FLAGS.validate:
         validate(datasets)
-    if FLAGS.update:
-        for dataset in datasets.values():
-            for reaction in dataset.reactions:
-                _update_reaction(reaction)
+    if not FLAGS.update:
+        logging.info('nothing else to do; use --update for more')
+        return  # Nothing else to do.
+    for dataset in datasets.values():
+        for reaction in dataset.reactions:
+            _update_reaction(reaction)
     # Combine all datasets into a single object.
     combined = None
-    for dataset in datasets.values():
+    for key in sorted(datasets):
+        dataset = datasets[key]
         if combined is None:
             combined = dataset
         else:
             combined.reactions.extend(dataset.reactions)
     if FLAGS.output:
-        message_helpers.write_message(
-            combined, FLAGS.output, FLAGS.input_format)
+        output_filename = FLAGS.output
+    else:
+        suffix = message_helpers.get_suffix(FLAGS.input_format)
+        dataset_id = uuid.uuid4().hex
+        output_filename = os.path.join(
+            'data', dataset_id[:2], f'{dataset_id}{suffix}')
```

Comment:
I think this is a good solution for organizing subfolders

Reply:
Yes; it was your idea :p